shared/sbl.js

```diff
@@ -0,0 +1,111 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+const DOMAIN_SELECTOR_REGEX = /^([^/]+)(.+)/
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = DOMAIN_SELECTOR_REGEX.exec(path)
+  var domain = domains[name]
+  // Filters can perform additional functions, and by returning `false` they
+  // can prevent the execution of a selector.
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain]
+  }
+}
+
+// During startup of app, use this instead of sbl function to call selectors
+// on domains that haven't been registered yet. The selector gets run immediately
+// once the corresponding domain is registered via registerDomain.
+sbl.ready = function (path: string, ...data?: Array<*>) {
+  // a call might happen to a selector for domains that haven't been registered
+  // yet, due to module load order. So we save them and run them later.
```

Comment:
Since we can't predict order it should check if the domain exists and execute if it does. Otherwise, it may never execute when registration has already happened.